src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +121,287 @@ object RuleUtils {
+      //   the appended files similar to indexPlan.
+      // Step 2) If Shuffle is required, perform shuffle for the plan.
+      // Step 3) Merge both indexPlan and planForAppended by using [[BucketUnion]] or [[Union]].
+      // For more details, see https://github.com/microsoft/hyperspace/issues/150.
+
+      val planForAppended =
+        transformPlanToReadFromAppendedFiles(spark, index.schema, plan, filesAppended)
+      if (useBucketSpec) {
+        // If Bucketing information of the index is used to read the index data, we need to
+        // shuffle the appended data in the same way to correctly merge with bucketed index data.
+
+        // Clear sortColumnNames as BucketUnion does not keep the sort order within a bucket.
+        val bucketSpec = index.bucketSpec.copy(sortColumnNames = Seq())
+
+        // Merge index plan & newly shuffled plan by using bucket-aware union.
+        BucketUnion(
+          Seq(indexPlan, transformPlanToShuffleUsingIndexSpec(bucketSpec, planForAppended)),
+          bucketSpec)
+      } else {
+        // If bucketing is not necessary (e.g. FilterIndexRule), we use [[Union]] to merge
+        // the appended data without additional shuffle.
+        Union(indexPlan, planForAppended)
+      }
+    } else {
+      indexPlan
+    }
+  }
+
+  /**
+   * Transform the current plan to read the given appended source files.
+   *
+   * The result will be merged with the plan which is reading index data
+   * by using [[BucketUnion]] or [[Union]].
+   *
+   * @param spark Spark session.
+   * @param indexSchema Index schema used for the output.
+   * @param originalPlan Original plan.
+   * @param filesAppended Appended files to the source relation.
+   * @return Transformed linear logical plan for appended files.
+   */
+  private def transformPlanToReadFromAppendedFiles(
+      spark: SparkSession,
+      indexSchema: StructType,
+      originalPlan: LogicalPlan,
+      filesAppended: Seq[Path]): LogicalPlan = {
+    // Replace the location of LogicalRelation with appended files.
+    val planForAppended = originalPlan transformDown {
+      case baseRelation @ LogicalRelation(fsRelation: HadoopFsRelation, baseOutput, _, _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion.
+        val updatedOutput =
+          baseOutput.filter(attr => indexSchema.fieldNames.contains(attr.name))
+        val newLocation =
+          new InMemoryFileIndex(spark, filesAppended, Map(), None)
+        val newRelation =
+          fsRelation.copy(
+            location = newLocation,
+            dataSchema = StructType(indexSchema.filter(baseRelation.schema.contains(_))))(spark)
+        baseRelation.copy(relation = newRelation, output = updatedOutput)
+    }
+    assert(!originalPlan.equals(planForAppended))
+    planForAppended
+  }
+
+  /**
+   * Transform the plan to perform on-the-fly Shuffle the data based on bucketSpec.
```

Comment:
I am a bit confused with "IndexSpec" in the function name. Did you mean `transformPlanToShuffleUsingBucketSpec`?